### DISCUSSION TITLE
Helix on wasm, full in browser demo

### DISCUSSION THREAD

**Discussion by makemeunsee** [helix-editor/helix#8588] [Show and tell]
Hi there,

Some [recent](https://github.com/helix-editor/helix/discussions/4706) [topics](https://github.com/helix-editor/helix/discussions/8359) made me want to have Helix in the browser, so I made it :grinning: 

**Here's the demo: https://makemeunsee.github.io/helix/demo/**

Many things don't work (and can't really), panics are likely around many corners, but overall it's Helix in a browser.

See the [crate](https://github.com/makemeunsee/helix/tree/wasm32/helix-web) for more technical details.

Would love some feedback or tips, on how to eventually integrate this into the baseline, how to re-enable some features, or anything :sweat_smile: 

![image](https://s6.gifyu.com/images/S6Gdp.gif)

**Comment by archseer** [comment]
This is really impressive! I thought we'd have to replace the whole frontend but it's nice we can just rely on xterm.js. Would you mind opening a draft PR so we can discuss specific parts of the code? e.g. the nucleo threadpool issue: @pascalkuthe could probably add a wasm workaround there

**Reply by pascalkuthe** [reply]
I am not really sure here.  I don't think its really possible to separate nucleo from rayou/multitherading. Without the ability to run multiple threads it just becomes pretty impossible to run performance intense stuff in the background without blocking UI.

There is  https://github.com/GoogleChromeLabs/wasm-bindgen-rayon but it sounds really hary. It sounds like without using nightly rust you can't event use atomics in wasm. Tokio should have similar issues so no idea how @makemeunsee even got that working. I guessing they are using a singlethreaded runtime? We don't follow tokio best practices in many places (way too much time beteewn await points/potential deadlocks on a single threaded runtime) so this will likely lead to a ton of UI issues.

AFAIK wasm threading just really isn't there yet (and the rust ecosystem isn't either).

**Reply by makemeunsee** [reply]
> Would you mind opening a draft PR so we can discuss specific parts of the code?

Sure! I'll do that later today.

> Tokio should have similar issues so no idea how @makemeunsee even got that working.

To be frank I didn't expect it to work, but somehow async e.g. polls work with**out** the tokio runtime. I think `wasm-bindgen-futures` is providing the (single thread) runtime instead, that's something I meant to dig further into.

There was a comment in another discussion about using web workers, but I havent looked at all how it could integrate.

**Reply by makemeunsee** [reply]
Here's the PR: https://github.com/helix-editor/helix/pull/8607
Happy to continue the technical discussion there!

**Comment by kirawi** [comment]
I think it would look better if `fira-code` (or any common console font) was used for the demo.

**Reply by makemeunsee** [reply]
good point, fixed

**Comment by makemeunsee** [comment]
The **[demo](https://makemeunsee.github.io/helix/demo)** has been improved in some notable ways:
- Better default font, thanks kirawi for the tip
- True color enabled, thanks archseer for the tip
- All themes of the main repo are now embedded and can be loaded, try `:theme `
- Buffers can now be written and read and are persisted in web local storage, using e.g. `:w xyz`, `:o xyz`
  This has been very crudely implemented, and is probably terrible security wise, so **use at your own risk**, think twice before storing any kind of private/secret data anywhere and particularly when using this demo.
- This enables editing, reloading and persisting the configuration, try `:config-open`, `:config-reload`

And I'll probably add support for Rust and toml syntax highlighting in the next days.

edit: fixed the most obvious bugs of the read/write web storage impl, embedded the tutorial and made it the default buffer
